src/Component/spec/Metadata/MetadataSpec.php

```diff
@@ -11,27 +11,157 @@
+        $this->getApplicationName()->shouldReturn('app');
+        $this->getName()->shouldReturn('product.with.dots');
+    }
+
+    function it_has_application_name(): void
+    {
+        $this->getApplicationName()->shouldReturn('app');
+    }
+
+    function it_has_resource_name(): void
+    {
+        $this->getName()->shouldReturn('product');
+    }
+
+    function it_humanizes_simple_names(): void
+    {
+        $this->getHumanizedName()->shouldReturn('product');
+    }
+
+    function it_humanizes_more_complex_names(): void
+    {
+        $this->beConstructedThrough('fromAliasAndConfiguration', ['app.product_option', ['driver' => 'doctrine/orm']]);
+
+        $this->getHumanizedName()->shouldReturn('product option');
+    }
+
+    function it_has_plural_resource_name(): void
+    {
+        $this->getPluralName()->shouldReturn('products');
+    }
+
+    function it_has_driver(): void
+    {
+        $this->getDriver()->shouldReturn('doctrine/orm');
+    }
+
+    function it_has_state_machine_component(): void
+    {
+        $this->getStateMachineComponent()->shouldReturn('symfony');
+    }
+
+    function it_has_templates_namespace(): void
+    {
+        $this->getTemplatesNamespace()->shouldReturn('product');
+    }
+
+    function it_has_access_to_specific_config_parameter(): void
+    {
+        $this->getParameter('driver')->shouldReturn('doctrine/orm');
+    }
+
+    function it_checks_if_specific_parameter_exists(): void
+    {
+        $this->hasParameter('foo')->shouldReturn(false);
+        $this->hasParameter('driver')->shouldReturn(true);
+    }
+
+    function it_throws_an_exception_when_parameter_does_not_exist(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('getParameter', ['foo'])
+        ;
+    }
+
+    function it_has_access_to_specific_classes(): void
+    {
+        $this->getClass('model')->shouldReturn('App\Model\Resource');
+    }
+
+    function it_throws_an_exception_when_class_does_not_exist(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('getClass', ['foo'])
+        ;
+    }
+
+    function it_checks_if_specific_class_exists(): void
+    {
+        $this->hasClass('bar')->shouldReturn(false);
+        $this->hasClass('model')->shouldReturn(true);
+    }
+
+    function it_generates_service_id(): void
     {
-        $this->shouldImplement(LegacyMetadataInterface::class);
+        $this->getServiceId('factory')->shouldReturn('app.factory.product');
+        $this->getServiceId('repository')->shouldReturn('app.repository.product');
+        $this->getServiceId('form.type')->shouldReturn('app.form.type.product');
     }
 
-    function it_should_be_an_alias_of_metadata(): void
+    function it_generates_permission_code(): void
     {
-        $this->shouldBeAnInstanceOf(NewMetadata::class);
+        $this->getPermissionCode('show')->shouldReturn('app.product.show');
```

Comment:
Shouldn't it be based on resource name?
```suggestion
        $this->getPermissionCode('show')->shouldReturn('app.product.with.dots.show');
```

Reply:
in the let method, the product has "'app.product'" alias.